src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Utilities/DisposableBag.cs

```diff
@@ -121,80 +71,39 @@ internal void AddDisposable(IDisposable? disposable)
+        /// Removes an object from the bag. If done before the bag is disposed, this will prevent
+        /// <paramref name="disposable"/> from being disposed along with the bag itself.
         /// </summary>
-        /// <param name="disposable">The value to remove. May be <c>null</c>.</param>
-        internal void RemoveDisposable(IDisposable? disposable)
+        /// <param name="disposable">The object to remove.</param>
+        public void Remove(IDisposable? disposable)
         {
             if (disposable == null)
             {
                 return;
             }
 
-            lock (this)
-            {
-                _disposables = _disposables.Remove(disposable);
-            }
+            ImmutableInterlocked.Update(
+                ref _disposables,
+                (set, item) => set?.Remove(item)!,
+                disposable);
         }
 
         /// <summary>
-        /// Registers a callback to be invoked when and if a token is canceled,
-        /// protecting against <see cref="ObjectDisposedException"/> in the event that the
-        /// <see cref="CancellationTokenSource"/> has already been disposed.
+        /// Implemented only to allow collection initialization of this type.
         /// </summary>
-        internal static CancellationTokenRegistration RegisterNoThrowOnDispose(Action callback, CancellationToken token)
-        {
-            try
-            {
-                return token.Register(callback);
-            }
-            catch (ObjectDisposedException)
-            {
-                // The CancellationTokenSource has already been disposed.  It rejected the register.
-                // But now we know the CancellationToken is in its final state (either cancelled or not).
-                // So simulate the right behavior by invoking the callback or not, based on whether it was
-                // already canceled.
-                if (token.IsCancellationRequested)
-                {
-                    callback();
-                }
-
-                return new CancellationTokenRegistration();
-            }
-        }
-
-        /// <summary>
-        /// Calls <see cref="IDisposable.Dispose"/> on all elements in a sequence,
-        /// allowing the sequence itself or elements inside it to be null.
-        /// </summary>
-        internal static void DisposeAllIfNotNull(IEnumerable<IDisposable?> sequence, bool cacheSequence = false)
-        {
-            if (sequence != null)
-            {
-                if (cacheSequence)
-                {
-                    // This makes us impervious to changes in the sequence generated by disposing elements of the sequence.
-                    sequence = sequence.ToList();
-                }
-
-                foreach (IDisposable? item in sequence)
-                {
-                    item?.Dispose();
-                }
-            }
-        }
+        IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
```

Comment:
Externally this looks a lot like a collection (Add and Remove methods), is named like a collection (Bag), and implemented IEnumerable, so this feels like a trap that doesn't really help any. It wouldn't be a big deal for this to just be the following, would it?

```suggestion
        IEnumerator IEnumerable.GetEnumerator() => _disposables.GetEnumerator();
```

Reply:
What should it do if the bag is disposed? `_disposables` would be null in such a case.

Reply:
`_disposables?.GetEnumerator() ?? Enumerator.Empty // you know what I mean :P` ?

It's probably not worth doing if its not trivial, but in 30 years if the .NET industry treats all enumerables with suspicion I reserve the right to say "I told you so" :P